### PR TITLE
Modify FL server

### PR DESCRIPTION
**Pull request by mmilenkoski**
This PR introduces my proposal on how the server would function. 

Basically, the peers would connect to the server with an id, and then sent weights for each epoch. Additionally, they will be able to ask for the averaged weights per epoch from the server. The server will only send the averaged weights when it has received the weights from all peers. The peers can wait for some time for the averaged weights to be available. If they are not available, they will continue with local training. Next step would be to implement the front-side, at which time we might need to modify the server a bit as well.

### CODE REVIEW

**Comment by martinjaggi** [comment]
ids are same as in DeAI, right?

**Comment by mmilenkoski** [comment]
> ids are same as in DeAI, right?

yes, we can use the same id generation function

**Comment by martinjaggi** [comment]
maybe it makes sense to attach to each gradient message some meta-data too:  id of the client, which timestep of model this grad was computed.
and make sure the server can continue a round even if some clients fail to answer. (both issues similar in deai)

**Comment by mmilenkoski** [comment]
> maybe it makes sense to attach to each gradient message some meta-data too: id of the client, which timestep of model this grad was computed.
> and make sure the server can continue a round even if some clients fail to answer. (both issues similar in deai)

by timestamp of model you mean epoch? We have metadata about the client id (as field in the json object), epoch and task (as parameters of the url)

About the server continuing if some data is not received - what would be the best way to do this? Wait some time x from the first or last message for the current epoch, or something else?

**Comment by martinjaggi** [comment]
On Thu, Sep 23, 2021 at 2:35 PM Martin Milenkoski ***@***.***>
wrote:

> maybe it makes sense to attach to each gradient message some meta-data
> too: id of the client, which timestep of model this grad was computed.
> and make sure the server can continue a round even if some clients fail to
> answer. (both issues similar in deai)
>
> by timestamp of model you mean epoch? We have metadata about the client id
> (as field in the json object), epoch and task (as parameters of the url)
>
communication round #

> About the server continuing if some data is not received - what would be
> the best way to do this? Wait some time x from the first or last message
> for the current epoch, or something else?
>
can have a threshold in the config. if >80% have answered, the server
aggregates and finishes the round

**Comment by Saipraneet** [comment]
> then sent weights for each epoch

This is a good starting point. In future we may want to allow different users to make different number of updates depending on their compute availability.


> by timestamp of model you mean epoch? 

Ideally both communication round and wall clock time when transmitting (say UTC). This would be useful to deal with "stale" gradients which arrive at the server very late.


> can have a threshold in the config. if >80% have answered

This is an excellent suggestion. In practice I think people use *both* the threshold and also a timeout.

**Comment by Saipraneet** [comment]
Can we add some functionality at the server level to do hyper-parameter search as well? This I think is quite useful and necessary.

Essentially, the server can start training with different hyperparameters on different clients, and keep track of this. We can use this to for e.g. figure out what learning rate to use etc.

**Comment by martinjaggi** [comment]
sure, let's make a separate issue for that, easier to distribute the work then

**Review by martinjaggi — APPROVED** [review]

**Comment by mmilenkoski** [comment]
Added the following things:
- renamed epoch to (communication) round
- added a peer threshold and set it 80%
- added a timestamp when weights were sent

> Can we add some functionality at the server level to do hyper-parameter search as well? This I think is quite useful and necessary.
> 
> Essentially, the server can start training with different hyperparameters on different clients, and keep track of this. We can use this to for e.g. figure out what learning rate to use etc.

Thanks for the suggestion, I opened issue #3 for this.
